### PR TITLE
Implements iter_mut for LruMap

### PR DESCRIPTION
**Pull request by anujmax**
Implements iter_mut for LruMap, we needed this to be able to mutate the values while iterating. Discussion here: 

https://github.com/subspace/subspace/pull/2736/files#r1595217002

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1509,6 +1520,43 @@ where
 
 impl<'a, K, V, L> ExactSizeIterator for Iter<'a, K, V, L> where L: Limiter<K, V> {}
 
+/// An mutable iterator for the LRU map, values are mutable.
+pub struct IterMut<'a, K, V, L>
+where
+    L: Limiter<K, V>,
+{
+    map: &'a RawTable<Entry<K, V, L::LinkType>>,
```
Comment (koute) [inline src/lib.rs]:
```suggestion
    map: &'a mut RawTable<Entry<K, V, L::LinkType>>,
```

File: src/lib.rs [hunk 2]
```diff
@@ -1509,6 +1520,43 @@ where
...
+    newest: L::LinkType,
+    remaining: usize,
+}
+
+impl<'a, K, V, L> Iterator for IterMut<'a, K, V, L>
```
Comment (koute) [inline src/lib.rs:1536]: Also implement `DoubleEndedIterator` and `ExactSizeIterator` like it is for `Iter`.

File: src/lib.rs [hunk 3]
```diff
@@ -2542,4 +2591,39 @@ mod tests {
...
     }
+
+    #[test]
+    fn iter_mut_works() {
+        let mut lru = LruMap::new(UnlimitedCompact);
+        assert!(lru.is_empty());
+        assert_eq!(lru.len(), 0);
+        assert_eq!(lru.iter_mut().collect::<Vec<_>>(), vec![]);
+
+        let vec = vec!["D", "C", "B", "A"];
```
Comment (koute) [inline src/lib.rs]: Using `Vec`s as values kinda overcomplicates this test. Can you just use normal integers as values? (And for modifying the values you could e.g. multiply only even values, or something like that.)

File: src/lib.rs [hunk 4]
```diff
@@ -1431,6 +1431,17 @@ where
...
     }
 
+    /// A mutable iterator over all of the elements in the most recently used order.
+    pub fn iter_mut(&mut self) -> IterMut<K, V, L> {
+        let newest = self.newest;
+        let len = self.len();
+        IterMut {
+            map: &mut self.map,
+            newest,
+            remaining: len,
```
Comment (koute) [inline src/lib.rs]: Stylistic nit.

```suggestion
        let newest = self.newest;
        let remaining = self.len();
        IterMut {
            map: &mut self.map,
            newest,
            remaining,
```

File: src/lib.rs [hunk 5]
```diff
@@ -1571,6 +1619,7 @@ mod tests {
     pub use super::*;
 
     extern crate std;
+    use core::usize;
```
Comment (koute) [inline src/lib.rs]: Why is this extra import here?

**Review by koute — CHANGES_REQUESTED** [review]

**Comment by anujmax** [comment]
Hi @koute , could you please help review again.

**Review by koute — APPROVED** [review]
Thanks!

**Comment by koute** [comment]
Published as 0.2.3

**Pull request merged**